### PR TITLE
Busbar change pre-processor

### PR DESCRIPTION
**Pull request by pet-mit**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the new behavior (if this is a feature change)?**
Read CRAC and create switches for BusBarChange RAs

### CODE REVIEW

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+    /**
+     * Move a branch's given side to a given bus
+     * (The method actually copies the branch to a new one then deletes it)
+     *
+     * @param branch the branch to move
+     * @param side the side of the branch to move
+     * @param bus  the new bus to connect to the side
+     * @return the new line
+     */
+    public Branch<?> moveBranch(Branch<?> branch, Branch.Side side, Bus bus) {
```
Comment (TheMaskedTurtle) [inline cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java:80]: This method could return void

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+
+    private static String getOrderCode(Branch<?> branch) {
+        if (branch.hasProperty(UcteConstants.ORDER_CODE)) {
+            return branch.getProperty(UcteConstants.ORDER_CODE);
+        } else {
+            return branch.getId().substring(branch.getId().length() - 1);
+        }
+    }
+
+    private Line moveLine(Line line, Branch.Side side, Bus bus) {
```
Comment (TheMaskedTurtle) [inline cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java:113]: Same

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+        setBranchAdderProperties(adder, line, side, bus);
+        Line newLine = adder.add();
+        copyCurrentLimits(line, newLine);
+        copyProperties(line, newLine);
+        LOGGER.debug("Line '{}' has been removed, new TieLine '{}' has been created", line.getId(), newLine.getId());
+        connectablesToRemove.add(line);
+        return newLine;
+    }
+
+    private Line moveTieLine(TieLine tieLine, Branch.Side side, Bus bus) {
```
Comment (TheMaskedTurtle) [inline cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java:125]: Same

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+        setBranchAdderProperties(adder, tieLine, side, bus);
+        TieLine newTieLine = adder.add();
+        copyCurrentLimits(tieLine, newTieLine);
+        copyProperties(tieLine, newTieLine);
+        LOGGER.debug("TieLine '{}' has been removed, new TieLine '{}' has been created", tieLine.getId(), newTieLine.getId());
+        connectablesToRemove.add(tieLine);
+        return newTieLine;
+    }
+
+    private TwoWindingsTransformer moveTwoWindingsTransformer(TwoWindingsTransformer transformer, Branch.Side side, Bus bus) {
```
Comment (TheMaskedTurtle) [inline cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java:137]: Same

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+                return adder.setVoltageLevel1(branch.getTerminal1().getVoltageLevel().getId())
+                    .setConnectableBus1(branch.getTerminal1().getBusBreakerView().getConnectableBus().getId())
+                    .setBus1(branch.getTerminal1().getBusBreakerView().getBus() != null ? branch.getTerminal1().getBusBreakerView().getBus().getId() : null);
+            } else if (side == Branch.Side.TWO) {
+                return adder.setVoltageLevel2(branch.getTerminal2().getVoltageLevel().getId())
+                    .setConnectableBus2(branch.getTerminal2().getBusBreakerView().getConnectableBus().getId())
+                    .setBus2(branch.getTerminal2().getBusBreakerView().getBus() != null ? branch.getTerminal2().getBusBreakerView().getBus().getId() : null);
+            }
+        }
+        throw new AssertionError();
```
Comment (TheMaskedTurtle) [inline cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java:205]: Throw FaraoException or just remove the if and the exception

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * This class exposes useful methods to modify a network
+ *
+ * @author Peter Mitri {@literal <peter.mitri@rte-france.com>}
+ */
+public class NetworkModifier {
```
Comment (TheMaskedTurtle) [inline cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java:23]: I think we should catch PowsyblException somewhere as well and report them properly without the process to crash

File: cse-lib/network-processing/src/test/java/com/farao_community/farao/cse/network_processing/TestUtils.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    public static void assertNetworksAreEqual(Network network, String reference) {
+        MemDataSource dataSource = new MemDataSource();
+
+        UcteExporter exporter = new UcteExporter();
+        exporter.export(network, new Properties(), dataSource);
+
+        try (InputStream actual = dataSource.newInputStream(null, "uct");
+             InputStream expected = TestUtils.class.getResourceAsStream(reference)) {
```
Comment (TheMaskedTurtle) [inline cse-lib/network-processing/src/test/java/com/farao_community/farao/cse/network_processing/TestUtils.java:32]: You can pass a class in arguments (Class<?> clazz) so that it will not be based on TestUtils that could be placed anywhere

File: cse-lib/network-processing/src/test/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessorTest.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class BusBarChangeProcessorTest {
+
+    private Network network;
+    private Set<BusBarChangeSwitches> busBarChangeSwitchesSet;
+
+    private void setUp(String networkFile, String cracFile) {
+        network = Importers.loadNetwork(networkFile, getClass().getResourceAsStream("/BusBarChange/" + networkFile), LocalComputationManager.getDefault(), new ImportConfig(), null);
```
Comment (TheMaskedTurtle) [inline cse-lib/network-processing/src/test/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessorTest.java:28]: You can just do 
`Importers.loadNetwork(networkFile, getClass().getResourceAsStream("/BusBarChange/" + networkFile))`

Comment (TheMaskedTurtle) [inline cse-lib/network-processing/src/test/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessorTest.java:28]: It would be cleaner to recreate package path in the resources com/farao_community/farao/cse/network_processing/busbar_change and place all the files there. Then you could access directly by getResourceAsStream(networkFile)

File: cse-lib/network-processing/src/test/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessorTest.java [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        InputStream is = getClass().getResourceAsStream("/BusBarChange/" + cracFile);
```
Comment (TheMaskedTurtle) [inline cse-lib/network-processing/src/test/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessorTest.java:29]: Same

File: pom.xml [hunk 10]
```diff
@@ -89,6 +91,11 @@
                 <artifactId>commons-io</artifactId>
                 <version>${commons.io.version}</version>
             </dependency>
+            <dependency>
+                <groupId>com.fasterxml.jackson.core</groupId>
+                <artifactId>jackson-databind</artifactId>
+                <version>${jackson.databind.version}</version>
+            </dependency>
```
Comment (TheMaskedTurtle) [inline pom.xml:98]: Why did you add that ?

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+
+    private Network network;
+    private UcteNetworkAnalyzer ucteNetworkAnalyzer;
+    private NetworkModifier networkModifier;
+
+    private Map<String, Set<SwitchPairToCreate>> switchesToCreatePerRa;
+    private Map<String, String> initialNodePerRa;
+    private Map<String, String> finalNodePerRa;
+    private Map<String, BusToCreate> busesToCreate;
+    private Map<String, Map<String, String>> createdSwitches;
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:49]: fixed

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+                triedIds += ", " + branchHelper.getUcteId();
+            }
+            if (!branchHelper.isValid()) {
+                throw new FaraoException(String.format("One of the branches in the remedial action was not found in the network (%s)", triedIds));
+            } else if (!(network.getIdentifiable(branchHelper.getIdInNetwork()) instanceof Line) && !(network.getIdentifiable(branchHelper.getIdInNetwork()) instanceof TwoWindingsTransformer)) {
+                throw new FaraoException(String.format("One of the branches (%s) in the remedial action is neither a line nor a two-windings-transformer: %s", branchHelper.getIdInNetwork(), network.getIdentifiable(branchHelper.getIdInNetwork()).getClass()));
+            }
+            if (!isBranchConnected(branchHelper.getIdInNetwork(), initialNodeId) && !isBranchConnected(branchHelper.getIdInNetwork(), finalNodeId)) {
+                throw new FaraoException(String.format("Branch %s is neither connected to initial node (%s) nor to final node (%s)", branchHelper.getIdInNetwork(), initialNodeId, finalNodeId));
+            }
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:148]: done

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+    }
+
+    /**
+     * For every BusBar TRemedialAction, this method detects and stores info about elements that should be created:
+     * - buses, if initial or final node does not exist in the network (info stored in busesToCreate)
+     * - switches, to be able to move the line from the initial to the final node (info stored in switchesToCreatePerRa)
+     */
+    private void computeBusesAndSwitchesToCreate(TRemedialAction tRemedialAction) {
+        String raId = tRemedialAction.getName().getV();
+        Set<BusToCreate> raBusesToCreate = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:97]: changed

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+            finalNodeId = finalNodeHelper.getIdInNetwork();
+            referenceBus = (Bus) network.getIdentifiable(finalNodeId);
+            initialNodeId = tRemedialAction.getBusBar().getInitialNode().getV();
+            raBusesToCreate.add(new BusToCreate(initialNodeId, referenceBus.getVoltageLevel().getId()));
+        } else {
+            throw new FaraoException(String.format("Remedial action's initial and final nodes are not valid: %s", initialNodeHelper.getInvalidReason()));
+        }
+
+        // Store all switches to create
+        switchesToCreatePerRa.put(raId, new HashSet<>());
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:125]: true

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java [hunk 15]
```diff
@@ -0,0 +1,330 @@
...
+        Set<BusToCreate> raBusesToCreate = new HashSet<>();
+        Set<SwitchPairToCreate> raSwitchesToCreate = new HashSet<>();
+
+        // Get initial and final nodes
+        UcteBusHelper initialNodeHelper = new UcteBusHelper(tRemedialAction.getBusBar().getInitialNode().getV(), ucteNetworkAnalyzer);
+        UcteBusHelper finalNodeHelper = new UcteBusHelper(tRemedialAction.getBusBar().getFinalNode().getV(), ucteNetworkAnalyzer);
+
+        String initialNodeId;
+        String finalNodeId;
+        Bus referenceBus;
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:106]: removed

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java [hunk 16]
```diff
@@ -0,0 +1,330 @@
...
+                throw new FaraoException(String.format("One of the branches in the remedial action was not found in the network (%s)", triedIds));
+            } else if (!(network.getIdentifiable(branchHelper.getIdInNetwork()) instanceof Line) && !(network.getIdentifiable(branchHelper.getIdInNetwork()) instanceof TwoWindingsTransformer)) {
+                throw new FaraoException(String.format("One of the branches (%s) in the remedial action is neither a line nor a two-windings-transformer: %s", branchHelper.getIdInNetwork(), network.getIdentifiable(branchHelper.getIdInNetwork()).getClass()));
+            }
+            if (!isBranchConnected(branchHelper.getIdInNetwork(), initialNodeId) && !isBranchConnected(branchHelper.getIdInNetwork(), finalNodeId)) {
+                throw new FaraoException(String.format("Branch %s is neither connected to initial node (%s) nor to final node (%s)", branchHelper.getIdInNetwork(), initialNodeId, finalNodeId));
+            }
+            SwitchPairToCreate switchPairToCreate = new SwitchPairToCreate(branchHelper.getIdInNetwork(), initialNodeId, finalNodeId);
+            raSwitchesToCreate.add(switchPairToCreate);
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:150]: changed

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java [hunk 17]
```diff
@@ -0,0 +1,330 @@
...
+        }
+        // else: should not happen, a check was done before
+
+        // Create switches
+        // Set OPEN/CLOSED status depending on the branch's initially connected node
+        Double currentLimit = getMinimumCurrentLimit(branch);
+        String switchOnInitial = networkModifier.createSwitch(voltageLevel, node1, fictitiousBus.getId(), currentLimit, !branchIsOnNode1).getId();
+        String switchOnFinal = networkModifier.createSwitch(voltageLevel, node2, fictitiousBus.getId(), currentLimit, branchIsOnNode1).getId();
+
+        return Pair.create(switchOnInitial, switchOnFinal);
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:258]: true

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java [hunk 18]
```diff
@@ -0,0 +1,330 @@
...
+    private String generateFictitiousBusId(VoltageLevel voltageLevel) {
+        char suffix = 'Z';
+        String busId;
+        do {
+            busId = String.format("%s%s", voltageLevel.getId(), suffix);
+            suffix -= 1;
+        }
+        while (network.getIdentifiable(busId) != null && suffix >= '9');
+        return busId;
+    }
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:286]: ok

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java [hunk 19]
```diff
@@ -0,0 +1,330 @@
...
+
+        SwitchPairToCreate(String branchId, String initialNodeId, String finalNodeId) {
+            this.branchId = branchId;
+            this.initialNodeId = initialNodeId;
+            this.finalNodeId = finalNodeId;
+            if (initialNodeId.compareTo(finalNodeId) < 0) {
+                this.uniqueId =  String.format("%s {%s, %s}", branchId, initialNodeId, finalNodeId);
+            } else {
+                this.uniqueId =  String.format("%s {%s, %s}", branchId, finalNodeId, initialNodeId);
+            }
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:322]: added

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java [hunk 20]
```diff
@@ -0,0 +1,330 @@
...
+
+        String getBusId() {
+            return busId;
+        }
+    }
+
+    /**
+     * Contains needed info to create a pair of switches
+     */
+    private static class SwitchPairToCreate {
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:308]: done

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java [hunk 21]
```diff
@@ -0,0 +1,330 @@
...
+            suffix -= 1;
+        }
+        while (network.getIdentifiable(busId) != null && suffix >= '9');
+        return busId;
+    }
+
+    /**
+     * Contains needed info to create a bus
+     */
+    private static class BusToCreate {
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:291]: done

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java [hunk 22]
```diff
@@ -0,0 +1,330 @@
...
+
+    /**
+     * Returns true if a line is connected to a given node
+     */
+    private boolean isBranchConnected(String branchId, String nodeId) {
+        Branch<?> branch = network.getBranch(branchId);
+        String bus1 = branch.getTerminal1().getBusBreakerView().getConnectableBus().getId();
+        String bus2 = branch.getTerminal2().getBusBreakerView().getConnectableBus().getId();
+        return bus1.equals(nodeId) || bus2.equals(nodeId);
+    }
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:168]: moved

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java [hunk 23]
```diff
@@ -0,0 +1,330 @@
...
+            Set<SwitchPairId> switchPairs = new HashSet<>();
+            switchPairsIds.forEach(switchPairId ->
+                switchPairs.add(
+                    new SwitchPairId(
+                        createdSwitches.get(switchPairId).get(initialNodePerRa.get(raId)),
+                        createdSwitches.get(switchPairId).get(finalNodePerRa.get(raId))
+                    )));
+            busBarChangeSwitches.add(new BusBarChangeSwitches(raId, switchPairs));
+        });
+        return busBarChangeSwitches;
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:212]: done, tell me if it's better

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java [hunk 24]
```diff
@@ -0,0 +1,316 @@
...
+    }
+
+    private static void findAndSetGeographicalName(Bus bus, VoltageLevel voltageLevel) {
+        for (Bus otherBus : voltageLevel.getBusBreakerView().getBuses()) {
+            if (otherBus.hasProperty(UcteConstants.GEOGRAPHICAL_NAME_PROPERTY_KEY)) {
+                bus.setProperty(UcteConstants.GEOGRAPHICAL_NAME_PROPERTY_KEY, otherBus.getProperty(UcteConstants.GEOGRAPHICAL_NAME_PROPERTY_KEY));
+                break;
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java:103]: just in case

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java [hunk 25]
```diff
@@ -0,0 +1,330 @@
...
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This class processes a CSE crac file in order to create, in the network, switches that can be used to apply
+ * BusBar remedial actions
+ *
+ * @author Peter Mitri {@literal <peter.mitri@rte-france.com>}
+ */
+public class BusBarChangeProcessor {
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:38]: done

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java [hunk 26]
```diff
@@ -0,0 +1,316 @@
...
+            .setB2(half2.getB2())
+            .setG1(half2.getG1())
+            .setG2(half2.getG2())
+            .setFictitious(half2.isFictitious())
+            .add()
+            .setUcteXnodeCode(xnodeCode);
+    }
+
+    private TwoWindingsTransformerAdder initializeTwoWindingsTransformerAdderToMove(TwoWindingsTransformer twoWindingsTransformer, String newId) {
+        return twoWindingsTransformer.getSubstation().orElseThrow().newTwoWindingsTransformer()
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java:283]: ok

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java (hunk 5 above)
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java:205]: ok

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java [hunk 27]
```diff
@@ -0,0 +1,316 @@
...
+            .setB1(line.getB1())
+            .setG2(line.getG2())
+            .setB2(line.getB2())
+            .setFictitious(line.isFictitious())
+            .setName(newId);
+    }
+
+    private static BranchAdder<?> setIdenticalToSide(Branch<?> branch, Branch.Side side, BranchAdder<?> adder) {
+        TopologyKind topologyKind = branch.getTerminal(side).getVoltageLevel().getTopologyKind();
+        if (topologyKind == TopologyKind.BUS_BREAKER) {
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java:194]: ok

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java (hunk 1 above)
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java:80]: changed

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java (hunk 2 above)
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java:113]: changed

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java (hunk 3 above)
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java:125]: changed

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java (hunk 4 above)
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java:137]: changed

File: cse-lib/network-processing/src/test/java/com/farao_community/farao/cse/network_processing/TestUtils.java (hunk 7 above)
Comment (pet-mit) [reply cse-lib/network-processing/src/test/java/com/farao_community/farao/cse/network_processing/TestUtils.java:32]: ok

File: cse-lib/network-processing/src/test/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessorTest.java (hunk 8 above)
Comment (pet-mit) [reply cse-lib/network-processing/src/test/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessorTest.java:28]: yes

Comment (pet-mit) [reply cse-lib/network-processing/src/test/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessorTest.java:28]: yes

File: cse-lib/network-processing/src/test/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessorTest.java (hunk 9 above)
Comment (pet-mit) [reply cse-lib/network-processing/src/test/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessorTest.java:29]: changed

File: pom.xml (hunk 10 above)
Comment (pet-mit) [reply pom.xml:98]: when i thought i could skip using the CSE crac importer and use raw xml. I forgot to remove it

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java (hunk 6 above)
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/NetworkModifier.java:23]: done in all public methods

File: cse-lib/network-processing/pom.xml [hunk 28]
```diff
@@ -13,6 +13,34 @@
...
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-ucte-converter</artifactId>
+        </dependency>
+
+        <!-- Runtime dependencies -->
+        <dependency>
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-iidm-impl</artifactId>
+            <scope>runtime</scope>
+        </dependency>
```
Comment (TheMaskedTurtle) [inline cse-lib/network-processing/pom.xml:36]: I think you don't need runtime dependency here, the client application would load it

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java [hunk 29]
```diff
@@ -0,0 +1,190 @@
...
+        return computeBusBarChangeSwitches(switchesToCreatePerRa, createdSwitches);
+    }
+
+    /**
+     * For every BusBar TRemedialAction, this method detects and stores info about elements that should be created:
+     * - buses, if initial or final node does not exist in the network (info stored in busesToCreate)
+     * - switches, to be able to move the line from the initial to the final node (info stored in switchesToCreatePerRa)
+     */
+    private static void computeBusesAndSwitchesToCreate(TRemedialAction tRemedialAction, UcteNetworkAnalyzer ucteNetworkAnalyzer, SortedSet<BusToCreate> busesToCreate, Map<String, Set<SwitchPairToCreate>> switchesToCreatePerRa) {
+        String raId = tRemedialAction.getName().getV();
```
Comment (TheMaskedTurtle) [inline cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:84]: Not necessary

**Review by TheMaskedTurtle — CHANGES_REQUESTED** [review]

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java (hunk 29 above)
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:84]: true

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java [hunk 30]
```diff
@@ -103,227 +91,99 @@ private void computeBusesAndSwitchesToCreate(TRemedialAction tRemedialAction) {
...
-                throw new FaraoException(String.format("One of the branches (%s) in the remedial action is neither a line nor a two-windings-transformer: %s", branchHelper.getIdInNetwork(), network.getIdentifiable(branchHelper.getIdInNetwork()).getClass()));
-            }
-            if (!isBranchConnected(branchHelper.getIdInNetwork(), initialNodeId) && !isBranchConnected(branchHelper.getIdInNetwork(), finalNodeId)) {
-                throw new FaraoException(String.format("Branch %s is neither connected to initial node (%s) nor to final node (%s)", branchHelper.getIdInNetwork(), initialNodeId, finalNodeId));
-            }
-            SwitchPairToCreate switchPairToCreate = new SwitchPairToCreate(branchHelper.getIdInNetwork(), initialNodeId, finalNodeId);
-            raSwitchesToCreate.add(switchPairToCreate);
-        }
+        tRemedialAction.getBusBar().getBranch().forEach(tBranch -> {
+            String linId = NetworkHelper.getLineIdInNetwork(tBranch.getFromNode().getV(), tBranch.getToNode().getV(), String.valueOf(tBranch.getOrder().getV()), initialNodeId, finalNodeId, ucteNetworkAnalyzer);
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:111]: I didn't want the NetworkHelper to depend on the crac object model

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java [hunk 31]
```diff
@@ -103,227 +91,99 @@ private void computeBusesAndSwitchesToCreate(TRemedialAction tRemedialAction) {
...
-                if (!createdSwitches.containsKey(switchPairToCreate.uniqueId)) {
-                    Pair<String, String> switches = createSwitchPair(switchPairToCreate);
-                    createdSwitches.put(switchPairToCreate.uniqueId,
-                        Map.of(switchPairToCreate.initialNodeId, switches.getFirst(),
-                            switchPairToCreate.finalNodeId, switches.getSecond()));
-                }
-            });
+    private static void createSwitches(Map<String, Set<SwitchPairToCreate>> switchesToCreatePerRa, Map<String, Map<String, String>> createdSwitches, NetworkModifier networkModifier) {
+        Set<SwitchPairToCreate> uniqueSwitchPairsToCreate = switchesToCreatePerRa.values().stream().flatMap(Set::stream).collect(Collectors.toSet());
+        uniqueSwitchPairsToCreate.stream().sorted().forEach(switchPairToCreate -> createdSwitches.put(switchPairToCreate.uniqueId, NetworkHelper.createSwitchPair(switchPairToCreate, networkModifier)));
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:133]: unfortunately, if I don't collect the items in a set, it wouldn't call the equals method to remove duplicates

[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:133]: (tested in testDoNotRecreateSwitches)

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java [hunk 32]
```diff
@@ -103,227 +91,99 @@ private void computeBusesAndSwitchesToCreate(TRemedialAction tRemedialAction) {
...
-                        createdSwitches.get(switchPairId).get(finalNodePerRa.get(raId))
-                    )));
+
+        // Loop through the RAs and add switch pairs IDs to the set
+        switchesToCreatePerRa.forEach((raId, switchPairsToCreateForRa) -> {
+            // Loop through the branches of RA and fetch switch pair IDs
+            Set<SwitchPairId> switchPairs = switchPairsToCreateForRa.stream().map(switchPairToCreate -> {
+                String switchPairId = switchPairToCreate.uniqueId();
+                // OPEN switch between branch and initial node, CLOSE switch between branch and final node
+                return new SwitchPairId(createdSwitches.get(switchPairId).get(switchPairToCreate.initialNodeId), createdSwitches.get(switchPairId).get(switchPairToCreate.finalNodeId));
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:148]: done

File: cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java (hunk 31 above)
[the remark this answers is not in the document]
Comment (TheMaskedTurtle) [reply cse-lib/network-processing/src/main/java/com/farao_community/farao/cse/network_processing/busbar_change/BusBarChangeProcessor.java:132]: ok

**Review by TheMaskedTurtle — APPROVED** [review]
